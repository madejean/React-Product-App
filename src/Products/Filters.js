import React from 'react';

class Filters extends React.Component {
  constructor(props){
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }
  handleChange(e) {
    const value = e.target[e.target.type === 'checkbox' ? 'checked' : 'value']
    const name = e.target.name;
    this.props.onFilter({
      [name] : value
    })
  }
  render() {
    return (
      <form className="Filters-container">
        <input
          type="text"
          placeholder="Search..."
          value={this.props.searchFilter}
          name="searchFilter"
          onChange={this.handleChange}
        />
        <p>
          <input
            type="checkbox"
            name="stockOnlyFitler"
            checked={this.props.stockOnlyFitler}
            onChange={this.handleChange}
          />
          Only Show products in stock
        </p>
      </form>
    );
  }
}

export default Filters;
