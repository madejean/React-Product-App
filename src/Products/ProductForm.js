import React from 'react';

const initialValues = {id: '', category: '', price: '', stocked: false, name: ''};

class ProductForm extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      product: Object.assign({}, initialValues),
      errors: {
        message: ""
      }
    }
    this.handleChange = this.handleChange.bind(this);
    this.handleSave = this.handleSave.bind(this);
  }
  handleChange(e){
    console.log(this.state.product)
    const name = e.target.name;
    const value = e.target.type === 'checkbox' ? e.target.checked : e.target.value;
    this.setState((prevState) => {
      prevState.product[name] = value;
      if(!prevState.product.name){
        var message = " Name is required !"
        return {
          errors: {
            message: message,
          }
        }
      } else {
        return {
          product: prevState.product,
          errors: {},
        }
      }
    })
  }
  handleSave(e){
    if(!this.state.product.name){
      e.preventDefault();
      return;
    } else {
      e.preventDefault();
      this.props.onSave(this.state.product);
      this.setState({
        product: Object.assign({}, initialValues),
        errors: {
          message: ""
        }
      })
    }
  }

  render() {
    return (
      <form>
        <h3>Enter a new product</h3>
          <p>
            <label>Name</label><br />
            <input className={this.state.errors.message === " Name is required !" ? 'error' : ''} type="text" name="name" onChange={this.handleChange} value={this.state.product.name} />
            <span style={{color: "red"}}>{this.state.errors.message}</span>
          </p>
          <p>
            <label>Category</label><br />
            <input type="text" name="category" onChange={this.handleChange} value={this.state.product.category} />
          </p>
          <p>
            <label>Price</label><br />
            <input type="text" name="price" onChange={this.handleChange} value={this.state.product.price} />
          </p>
          <p>
            <label>In stock?</label><br />
            <input type="checkbox" name="stocked" onChange={this.handleChange} checked={this.state.product.stocked}/>
          </p>
          <input type="submit" value="save" onClick={this.handleSave} />
      </form>
    );
  }
}

export default ProductForm;
