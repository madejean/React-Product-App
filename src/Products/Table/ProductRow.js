import React from 'react';

class ProductRow extends React.Component {
  constructor(props){
    super(props);
    this.delete = this.delete.bind(this);
  }
  delete(){
    this.props.onDelete(this.props.product.id);
  }
  render() {
    var name = this.props.product.stocked ? this.props.product.name :
      <span style={{color: 'red'}}>{this.props.product.name}</span>
    return (
      <tr>
        <td>{name}</td>
        <td>{this.props.product.price}</td>
        <td><button onClick={this.delete}>x</button></td>
      </tr>
    );
  }
}

export default ProductRow;
