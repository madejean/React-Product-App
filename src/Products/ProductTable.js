import React from 'react';
import ProductRow from './Table/ProductRow.js';
import ProductTableHeader from './Table/ProductTableHeader.js';

class ProductTable extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      sort: {
        columnHeader: 'name',
        direction: 'desc'
      }
    };
    this.sortByColumnAndDirection = this.sortByColumnAndDirection.bind(this);
    this.handleSort = this.handleSort.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
  }
  sortByColumnAndDirection(objectA, objectB){
    //checkout javascript sorting api
    let isDesc = this.state.sort.direction === 'desc' ? 1 : -1;
    let [a, b] = [objectA[this.state.sort.columnHeader], objectB[this.state.sort.columnHeader]];
    if(this.state.sort.column === 'price') {
      [a, b] = [a, b].map((val) => parseFloat(val.replace(/[^\d.]/g, ''), ''), 10);
    }
    if (a > b) {
      return 1 * isDesc;
    }
    if (a < b) {
      return -1 * isDesc;
    }
    return 0;
  }
  sortProducts(){
    var productsArray = Object.keys(this.props.products).map((p) => this.props.products[p]);
    return productsArray.sort(this.sortByColumnAndDirection);
  }
  handleSort(columnHeader, direction){
    this.setState({
      sort: {
        columnHeader: columnHeader,
        direction: direction
      }
    })
  }
  handleDelete(id){
    this.props.onDelete(id)
  }
  render() {
    var rows = [];
    this.sortProducts().forEach((product) => {
      if (product.name.indexOf(this.props.searchFilter) === -1 || (!product.stocked && this.props.stockOnlyFitler)){
        return;
      }
      rows.push(
        <ProductRow product={product} key={product.id} onDelete={this.handleDelete} />
      );
    });
    return (
      <div>
        <table>
          <thead>
            <tr>
              <ProductTableHeader
                onSort={this.handleSort}
                currentSortOption={this.state.sort}
                columnHeader="name"
              />
              <ProductTableHeader
                onSort={this.handleSort}
                currentSortOption={this.state.sort}
                columnHeader="price"
              />
            </tr>
          </thead>
          <tbody>{rows}</tbody>
        </table>
      </div>
    );
  }
}

export default ProductTable;
