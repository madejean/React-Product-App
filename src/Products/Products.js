import React from 'react';
import Filters from './Filters.js';
import ProductForm from './ProductForm';
import ProductTable from './ProductTable.js';

export var PRODUCTS = {
  '1': {id: 1, category: 'Musical Instruments', price: '$459.99', stocked: true, name: 'Piano'},
  '2': {id: 2, category: 'Musical Instruments', price: '$5,000', stocked: true, name: 'Violon'},
  '3': {id: 3, category: 'Furniture', price: '$11,000', stocked: false, name: 'Sofa'},
  '4': {id: 4, category: 'Furniture', price: '$799', stocked: true, name: 'Dresser'},
  '5': {id: 5, category: 'Furniture', price: '$1,300', stocked: false, name: 'Dining Table'},
  '6': {id: 6, category: 'Furniture', price: '$300', stocked: true, name: 'TV'}
};

class Products extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      products: PRODUCTS,
      searchFilter: '',
      stockOnlyFitler: false
    }
    this.handleFilter = this.handleFilter.bind(this);
    this.save = this.save.bind(this);
    this.delete = this.delete.bind(this);
  }
  handleFilter(filterInput){
    this.setState(filterInput);
  }
  save(product){
    this.setState((prevState) => {
      var products = prevState.products;
      products[product.id] = product;
      return {products};
    })
  }
  delete(productId){
    this.setState((prevState) => {
      var products = prevState.products;
      delete products[productId];
      return { products };
    })
  }
  render() {
    return (
      <div>
        <Filters
          searchFilter={this.state.searchFilter}
          stockOnlyFitler={this.state.stockOnlyFitler}
          onFilter={this.handleFilter}
        />
        <ProductTable
          products={PRODUCTS}
          searchFilter={this.state.searchFilter}
          stockOnlyFitler={this.state.stockOnlyFitler}
          onDelete={this.delete}
        />
        <ProductForm
          onSave={this.save}
        />
      </div>
    );
  }
}

export default Products;
