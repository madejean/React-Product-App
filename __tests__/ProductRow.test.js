import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ProductRow from '../src/Products/Table/ProductRow';
import { PRODUCTS } from '../src/Products/Products';


test('ProductRow component should render as expected', () => {
  var product = {
    '1': {id: 1, category: 'Musical Instruments', price: '$459.99', stocked: true, name: 'Piano'},
  }
  const component = shallow(<ProductRow product={product}/>)
  const tree = toJson(component)
  expect(tree).toMatchSnapshot()
})
