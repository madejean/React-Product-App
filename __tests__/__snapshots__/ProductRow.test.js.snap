exports[`test ProductRow component should render as expected 1`] = `
<tr>
  <td>
    <span
      style={
        Object {
          "color": "red",
        }
      } />
  </td>
  <td />
  <td>
    <button
      onClick={[Function]}>
      x
    </button>
  </td>
</tr>
`;
